### PR TITLE
refactor(pgwire): unify query logs and slow query logs

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Provide the common info like `mode`, `session`, `sql` with a tracing span in the entrance function `do_process`, which...

- avoid code duplication on different message types
- is reused with #10660
- unify the path for query logs and slow query logs

The slow query log was introduced in #6624 a long time ago. However, it has recently been discovered (in https://github.com/risingwavelabs/risingwave/pull/13287#discussion_r1386357091 and also by @liurenjie1024) that it fails to take into account the time required to consume the data stream. This PR also fixes that by handling all these stuff in `pgwire`.

Tested locally:
- `query.log`
```
2023-12-12T16:43:36.017757+08:00 risingwave-main ThreadId(10) handle_query{mode="simple query" session_id=0 sql=select fu;}: status="err" time=68.853375ms
2023-12-12T16:45:20.525747+08:00 risingwave-main ThreadId(06) handle_query{mode="simple query" session_id=0 sql=select pg_sleep(65);}: status="ok" time=65.020061167s
```
- `slow_query.log`
```
2023-12-12T16:45:15.508607+08:00 risingwave-main ThreadId(07) handle_query{mode="simple query" session_id=0 sql=select pg_sleep(65);}: slow query elapsed=60s
```

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by liurenjie1024 — APPROVED** [review]
LGTM, thanks!

**Review by chenzl25 — COMMENTED** [review]
LGTM. Good Catch

File: src/common/src/util/query_log.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+/// The target name for the root span while pgwire is handling a query.
+pub const PGWIRE_ROOT_SPAN_TARGET: &str = "pgwire_root_span";
+/// The target name for the event when pgwire finishes handling a query.
+pub const PGWIRE_QUERY_LOG: &str = "pgwire_query_log";
+/// The target name for the event when pgwire does not finish handling a query in time.
+pub const PGWIRE_SLOW_QUERY_LOG: &str = "pgwire_slow_query_log";
+
+/// The period of logging ongoing slow queries.
+pub static SLOW_QUERY_LOG_PERIOD: LazyLock<Duration> = LazyLock::new(|| {
```
[the remark this answers is not in the document]
Comment (xxchan) [reply src/common/src/util/query_log.rs:26]: https://github.com/risingwavelabs/risingwave/issues/13961 created an issue

File: src/utils/runtime/src/logger.rs [hunk 2]
```diff
@@ -247,53 +247,73 @@ pub fn init_risingwave_logger(settings: LoggerSettings) {
...
+        /// will cause several problems:
+        ///
+        /// - `with_ansi(false)` does not take effect and it will follow the settings of
+        ///   the primary fmt layer installed above.
+        /// - `Span::record` will update the same `FormattedFields` multiple times,
+        ///   leading to duplicated fields.
+        ///
+        /// As a workaround, we use a newtype wrapper here to get a different type id.
+        /// The const generic parameter `SLOW` is further used to distinguish between the
+        /// query log and the slow query log.
```
Comment (xxchan) [inline src/utils/runtime/src/logger.rs:264]: Specifically, what part of the following log is controlled by this?
```
2023-12-12T16:43:36.017757+08:00 risingwave-main ThreadId(10) handle_query{mode="simple query" session_id=0 sql=select fu;}: status="err" time=68.853375ms

```

File: src/utils/pgwire/src/pg_protocol.rs [hunk 3]
```diff
@@ -199,37 +208,122 @@ where
...
+    /// [`PGWIRE_ROOT_SPAN_TARGET`].
+    ///
+    /// This is used to provide context for the (slow) query logs and traces.
+    ///
+    /// The span is only effective if there's a current session and the message is
+    /// query-related. Otherwise, `Span::none()` is returned.
+    fn root_span_for_msg(&self, msg: &FeMessage) -> tracing::Span {
+        let Some(session_id) = self.session.as_ref().map(|s| s.id().0) else {
+            return tracing::Span::none();
+        };
```
Comment (xxchan) [inline src/utils/pgwire/src/pg_protocol.rs:221]: When is there no session? :eyes:

File: src/utils/pgwire/src/pg_protocol.rs [hunk 4]
```diff
@@ -460,26 +554,12 @@ where
     async fn process_query_msg(&mut self, query_string: io::Result<&str>) -> PsqlResult<()> {
         let sql: Arc<str> =
             Arc::from(query_string.map_err(|err| PsqlError::SimpleQueryError(Box::new(err)))?);
-        let start = Instant::now();
+        record_sql_in_span(&sql);
```
Comment (xxchan) [inline src/utils/pgwire/src/pg_protocol.rs:557]: a little confused. Does this duplicate with the layer in `do_process`?

Comment (BugenZhao) [reply src/utils/pgwire/src/pg_protocol.rs:557]: > the layer in do_process?

What do you mean by this? 😳

File: src/utils/pgwire/src/pg_protocol.rs (hunk 3 above)
Comment (BugenZhao) [reply src/utils/pgwire/src/pg_protocol.rs:221]: I guess when the `Startup` message is not handled?

File: src/utils/pgwire/src/pg_protocol.rs (hunk 4 above)
Comment (xxchan) [reply src/utils/pgwire/src/pg_protocol.rs:557]: Oh, `do_process` emits the log line, but `process_query_msg` adds the sql field. I see.

File: src/utils/runtime/src/logger.rs [hunk 5]
```diff
@@ -247,53 +247,73 @@ pub fn init_risingwave_logger(settings: LoggerSettings) {
...
+        #[derive(Default)]
+        struct FmtFields<const SLOW: bool>(DefaultFields);
```
Comment (fuyufjh) [inline src/utils/runtime/src/logger.rs:266]: Where is the `SLOW` used?

File: src/utils/pgwire/src/pg_protocol.rs [hunk 6]
```diff
@@ -199,37 +208,122 @@ where
...
+            }
 
-        match result {
+            // Log to optionally-enabled target `PGWIRE_QUERY_LOG`.
+            // Only log if we're currently in a tracing span set in `span_for_msg`.
+            if !tracing::Span::current().is_none() {
+                tracing::info!(
+                    target: PGWIRE_QUERY_LOG,
+                    status = if result.is_ok() { "ok" } else { "err" },
+                    time = ?elapsed,
```
Comment (chenzl25) [inline src/utils/pgwire/src/pg_protocol.rs:315]: BTW, shall we print the `elapsed` in a fixed width and fixed unit format?

File: src/utils/pgwire/src/pg_protocol.rs [hunk 7]
```diff
@@ -160,6 +158,17 @@ pub fn cstr_to_str(b: &Bytes) -> Result<&str, Utf8Error> {
     std::str::from_utf8(without_null)
 }
 
+/// Record `sql` in the current tracing span.
+fn record_sql_in_span(sql: &str) {
```
Comment (fuyufjh) [inline src/utils/pgwire/src/pg_protocol.rs:162]: What's the relationship between this and the `run_stmt` in #10660?

**Review by fuyufjh — COMMENTED** [review]
Overall, I feel that "unifying query logs and slow query logs" doesn't seem to reduce the complexity but somehow make it more complicated to me... Tend to keep the original implementation, which was simple and straight-forward.

File: src/utils/pgwire/src/pg_protocol.rs [hunk 8]
```diff
@@ -199,37 +208,122 @@ where
...
+                            ?elapsed,
+                            "slow query"
+                        );
+                    }
+                }
+            }
+        };
+
+        // Query log.
+        let fut = async move {
```
Comment (xxchan) [inline src/utils/pgwire/src/pg_protocol.rs:299]: Do the consecutive `async move` layers have performance implications? 🤔

File: src/utils/pgwire/src/pg_protocol.rs (hunk 7 above)
Comment (BugenZhao) [reply src/utils/pgwire/src/pg_protocol.rs:162]: I assume you're talking about the span created here:

https://github.com/risingwavelabs/risingwave/blob/9dcad747a462f9dd1df571f8812ffa5f36596197/src/utils/pgwire/src/pg_protocol.rs#L230-L236

This replaces the old one, which was renamed to `process_query_msg_one_stmt` after that PR. Pros:

- It's defined in the unified path for handling `Query` `Parse` and `Execute`, while the old one only handled (simple) `Query`.
- It's now not only used by traces, but also used to provide common info for the query logs later, including the slow ones.
https://github.com/risingwavelabs/risingwave/blob/9dcad747a462f9dd1df571f8812ffa5f36596197/src/utils/pgwire/src/pg_protocol.rs#L312-L316
https://github.com/risingwavelabs/risingwave/blob/9dcad747a462f9dd1df571f8812ffa5f36596197/src/utils/pgwire/src/pg_protocol.rs#L288-L292

As for recording the SQL field after creating the span: I just don't want to duplicate the code for extracting the SQL from the messages. This also reduces allocations.

File: src/utils/pgwire/src/pg_protocol.rs (hunk 8 above)
Comment (BugenZhao) [reply src/utils/pgwire/src/pg_protocol.rs:299]: I guess not intuitively. 🤣 This is just like the adaptors for iterators or futures.

File: src/utils/runtime/src/logger.rs (hunk 2 above)
Comment (BugenZhao) [reply src/utils/runtime/src/logger.rs:264]:
```diff
- handle_query
+ <some control character for ANSI color>handle_query
- {mode="simple query" session_id=0 sql=select fu;}:
+ {mode="simple query" session_id=0 sql=select fu; sql=select fu; sql=select fu;}:
 status="err" time=68.853375ms
```

File: src/utils/runtime/src/logger.rs (hunk 5 above)
Comment (BugenZhao) [reply src/utils/runtime/src/logger.rs:266]: https://github.com/risingwavelabs/risingwave/blob/9dcad747a462f9dd1df571f8812ffa5f36596197/src/utils/runtime/src/logger.rs#L303-L306

There's no actual use for this flag. It just makes the compiler to treat them as different types and then assign them with different [`type_id`](https://doc.rust-lang.org/stable/std/any/trait.Any.html#tymethod.type_id)s, which is used for RTII.

File: src/utils/pgwire/src/pg_protocol.rs (hunk 6 above)
Comment (BugenZhao) [reply src/utils/pgwire/src/pg_protocol.rs:315]: Can you elaborate more? I'm afraid it's hard to do this for the duration from `50 µs = 0.05 ms` to `5 min = 300 000 ms`. 🥺

File: src/utils/runtime/src/logger.rs (hunk 2 above)
Comment (BugenZhao) [reply src/utils/runtime/src/logger.rs:264]: Why it's colorful: https://docs.rs/tracing-subscriber/0.3.18/src/tracing_subscriber/fmt/fmt_layer.rs.html#815-823

Why `sql` is recorded multiple times: https://docs.rs/tracing-subscriber/0.3.18/src/tracing_subscriber/fmt/fmt_layer.rs.html#851-854

Notice the type parameter `N` here.

**Comment by BugenZhao** [comment]
> Overall, I feel that "unifying query logs and slow query logs" doesn't seem to reduce the complexity but somehow make it more complicated to me... Tend to keep the original implementation, which was simple and straight-forward.

To be honest, this PR is more like a fix and a feature than a refactor.

- Frontend tracing now works in extended mode as well.
- The slow query log now correctly records the time by moving it to pgwire.

If the refactor were not done, then the code for providing the common fields like `sql` or `session_id` has to be repeated 3 (log, slow log, trace span) x 3(query, parse, execute) = 9 times, also including the code for truncating and possibly redacting (#12673). I believe DRY is necessary here. 😕

File: src/utils/pgwire/src/pg_protocol.rs (hunk 6 above)
Comment (chenzl25) [reply src/utils/pgwire/src/pg_protocol.rs:315]: I think `ms` is enough for batch query. How about using `.as_millis()`? With the same time unit could let us easy to grep these lines.

**Pull request merged**